website/blog/2024-12-04-from-natural-language-to-sql/index.md

```diff
@@ -0,0 +1,1067 @@
+
+3. **Cyclical Graph Structure**: The graph’s cyclical nature introduces a robust retry mechanism. This mechanism dynamically addresses failures by looping back to previous stages when needed, ensuring continuous graph execution. (Details of this mechanism will be discussed later.)
+## AI Workflow Overview
+
+The Multilingual Query Engine’s advanced AI workflow is composed of interconnected nodes and edges, each representing a crucial stage:
+
+1. **Translation Node**: Converts the user’s input into English.
+
+2. **Safety Checks**: Ensures user input is free from toxic or inappropriate content and does not contain harmful SQL commands (e.g., DELETE, DROP).
+
+3. **Database Schema Extraction**: Retrieves the schema of the target database to understand its structure and available data.
+
+4. **Relevancy Validation**: Validates the user’s input against the database schema to ensure alignment with the database’s capabilities.
+
+5. **SQL Query Generation**: Generates an SQL query based on the user’s input and the current database schema.
+
+6. **SQL Query Validation**: Executes the SQL query in a rollback-safe environment to ensure its validity before running it.
+
+7. **Dynamic State Evaluation**: Determines the next steps based on the current state. If the SQL query validation fails, it loops back to Stage 5 to regenerate the query.
+
+8. **Query Execution and Result Retrieval**: Executes the SQL query and returns the results if it’s a SELECT statement.
+
+The retry mechanism is introduced in Stage 7, where the system dynamically evaluates the current graph state. Specifically, when the SQL query validation node (Stage 6) detects an issue, the state triggers a loop back to the SQL Generation node (Stage 5) for a new SQL Generation attempt (within a maximum of 3 attemps).
+
+## Components
+
+The Multilingual Query Engine interacts with several external components to transform Natural Language user inputs into SQL queries and execute them in a safe and robust manner. In this section, we will take a detailed look at the key AI Workflow components: OpenAI, Vector Store, SQLite Database, and SQL Generation Chain.
+
+### OpenAI
+
+OpenAI, more specifically gpt-4 language model plays a crucial role in multiple stages of the workflow, providing the intelligence required for:
+
+1. **Translation**: Translates user input into English. If the text is already in English, it simply repeats the input.
+
+2. **Safety Checks**: Analyzes user input to ensure that it does not contain toxic or inappropriate content.
+
+3. **Relevancy Checks**: Evaluates whether the user's question is relevant given the database schema.
+
+4. **SQL Generation**: Generates valid and executable SQL queries based on user input, SQL generation documentation, and the database schema.
+
+Details on OpenAI implementation will be provided later on in the Node implementation section.
+### FAISS Vector Store
+
```

Comment:
Might want to mention alternatives to an in-memory Vector Store for permanent / more scalable shared resource embeddings storage that can be shared across projects.